internal/controllers/clusterextension_controller.go

```diff
@@ -112,88 +179,244 @@ func checkForUnexpectedFieldChange(a, b ocv1alpha1.ClusterExtension) bool {
+				ensureAllConditionsWithReason(ext, "ResolutionFailed", errorMessage)
+			} else {
+				// General error handling
+				setResolvedStatusConditionFailed(&ext.Status.Conditions, errorMessage, ext.Generation)
+				ensureAllConditionsWithReason(ext, "InstallationStatusUnknown", "")
+			}
+		}
+	} else {
+		// If the error is not an aggregate, handle it as a general error
+		errorMessage := err.Error()
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, errorMessage, ext.Generation)
+		ensureAllConditionsWithReason(ext, "InstallationStatusUnknown", "")
+	}
+	ext.Status.ResolvedBundle = nil
+	return ctrl.Result{}, err
+}
+
 // Helper function to do the actual reconcile
 //
 // Today we always return ctrl.Result{} and an error.
 // But in the future we might update this function
 // to return different results (e.g. requeue).
 //
+/* The reconcile functions performs the following major tasks:
+1. Resolution: Run the resolution to find the bundle from the catalog which needs to be installed.
+2. Validate: Ensure that the bundle returned from the resolution for install meets our requirements.
+3. Unpack: Unpack the contents from the bundle and store in a localdir in the pod.
+4. Install: The process of installing involves:
+4.1 Converting the CSV in the bundle into a set of plain k8s objects.
+4.2 Generating a chart from k8s objects.
+4.3 Apply the release on cluster.
+*/
 //nolint:unparam
 func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alpha1.ClusterExtension) (ctrl.Result, error) {
-	// Lookup the bundle that corresponds to the ClusterExtension's desired package.
-	bundle, err := r.resolve(ctx, ext)
+	l := log.FromContext(ctx).WithName("operator-controller")
+	// run resolution
+	bundle, err := r.resolve(ctx, *ext)
 	if err != nil {
-		ext.Status.InstalledBundle = nil
-		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
-		ext.Status.ResolvedBundle = nil
-		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		l.V(1).Info("bundle resolve error", "error", err)
+		return r.handleResolutionErrors(ext, err)
+	}
+
+	if err := r.validateBundle(bundle); err != nil {
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as installation has failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
 
-		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+	bundleVersion, err := bundle.Version()
+	if err != nil {
+		ext.Status.ResolvedBundle = nil
+		ext.Status.InstalledBundle = nil
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonResolutionFailed, err), ext.GetGeneration())
```

Comment:
Nit: just put the error message here. The condition in the status should already have the appropriate reason, no need to repeat it in the message.

Reply:
Fixed in https://github.com/operator-framework/operator-controller/pull/887